### PR TITLE
more readable code (no templates) + a small addition to sol

### PR DESCRIPTION
**Pull request by devo1ution**
_If any of the below don't apply to this Pull Request, mark the checkbox as done._

- [x] I have tested my code.
- [x] I have added my solution according to the steps [here](https://usaco.guide/general/adding-solution#steps).
- [x] I have followed the code conventions mentioned [here](https://usaco.guide/general/adding-solution/#code-conventions).
  - I understand that if it is clear that I have not attempted to follow these conventions, my PR will be closed.
  - If changes are requested, I will re-request a review after addressing them.

### CODE REVIEW

File: solutions/gold/cses-1665.mdx [hunk 1]
```diff
@@ -40,375 +40,53 @@ information and problems.
...
-	str res = "{";
-	for (const auto &x : v) {
-		if (!fst) res += ", ";
-		fst = 0;
-		res += ts(x);
+const int N = 101, X = 5e3 + 1, M = 1e9 + 7,
+          K = 5e3;  // K is offset to account for negatives
+int n, x, t[N];     // x -> max penalty
+// dp[N][N][X] -> dp[i][j][k] = first i people, j unfinished groups, k penalty
+// contrary to the above solution, this code uses a push dp approach
```
Comment (SansPapyrus683) [inline solutions/gold/cses-1665.mdx:49]: > this code uses a push dp approach

it's very bad to do an editorial saying one method and then have code demonstrating another

Comment (devo1ution) [reply solutions/gold/cses-1665.mdx]: Thanks for the comment; I changed the sol to use push dp as well.

File: solutions/gold/cses-1665.mdx [hunk 2]
```diff
@@ -10,405 +10,70 @@ as $(\text{Skill of last person}) - (\text{Skill of first person})$. Let $s_i$
...
   we store (to save memory).
 
-I believe that this is called "open and close interval trick". See
+I believe that this is called the "open and close interval trick". See
 [zscoder's blog post](https://codeforces.com/blog/entry/47764) for more
 information and problems.
 
 ```cpp
 #include <bits/stdc++.h>
+#define int long long
```
Comment (SansPapyrus683) [inline solutions/gold/cses-1665.mdx:38]: use a `using` declaration instead & use `ll`

File: solutions/gold/cses-1665.mdx [hunk 3]
```diff
@@ -10,405 +10,70 @@ as $(\text{Skill of last person}) - (\text{Skill of first person})$. Let $s_i$
...
-tcTU > str ts(pair<T, U> p);
-tcT > str ts(T v) {  // containers with begin(), end()
-#ifdef LOCAL
-	bool fst = 1;
-	str res = "{";
-	for (const auto &x : v) {
-		if (!fst) res += ", ";
-		fst = 0;
-		res += ts(x);
+const int N = 101, X = 5e3 + 1, M = 1e9 + 7,
```
Comment (SansPapyrus683) [inline solutions/gold/cses-1665.mdx:40]: declarations on separate lines

File: solutions/gold/cses-1665.mdx [hunk 4]
```diff
@@ -10,405 +10,70 @@ as $(\text{Skill of last person}) - (\text{Skill of first person})$. Let $s_i$
...
+          K = 5e3;  // K is offset to account for negatives
+int n, x, s[N];     // x -> max penalty
+// dp[N][N][X] -> dp[i][j][k] = first i people, j unfinished groups, k penalty
+// for convenience, we 1-index dp array and 0-index person array so dp[i][j][k]
+// lines up with s[i] (i.e. skill of person i + 1 = s[i])
+int dp1[N][X + K], dp2[N][X + K];  // dp1 -> i, dp2 -> i + 1
+signed main() {
```
Comment (SansPapyrus683) [inline solutions/gold/cses-1665.mdx:47]:
```suggestion
int main() {
```

File: solutions/gold/cses-1665.mdx [hunk 5]
```diff
@@ -10,405 +10,70 @@ as $(\text{Skill of last person}) - (\text{Skill of first person})$. Let $s_i$
...
-		if (!fst) res += ", ";
-		fst = 0;
-		res += ts(x);
+const int N = 101, X = 5e3 + 1, M = 1e9 + 7,
+          K = 5e3;  // K is offset to account for negatives
+int n, x, s[N];     // x -> max penalty
+// dp[N][N][X] -> dp[i][j][k] = first i people, j unfinished groups, k penalty
+// for convenience, we 1-index dp array and 0-index person array so dp[i][j][k]
+// lines up with s[i] (i.e. skill of person i + 1 = s[i])
+int dp1[N][X + K], dp2[N][X + K];  // dp1 -> i, dp2 -> i + 1
```
Comment (SansPapyrus683) [inline solutions/gold/cses-1665.mdx:46]: can you just use vectors instead

File: solutions/gold/cses-1665.mdx [hunk 6]
```diff
@@ -10,405 +10,70 @@ as $(\text{Skill of last person}) - (\text{Skill of first person})$. Let $s_i$
...
-	dp[0][5000] = 1;
-	// ps(t);
-	trav(a, t) process(a);
-	mi ret(0);
-	FOR(i, 5000, 5000 + x + 1) ret += dp[0][i];
-	ps(ret);
-	// you should actually read the stuff at the bottom
+	int ans = 0;
+	for (int i = K; i <= x + K; i++)
+		ans += dp1[0][i],
```
Comment (SansPapyrus683) [inline solutions/gold/cses-1665.mdx:74]: braces

File: solutions/gold/cses-1665.mdx [hunk 7]
```diff
@@ -10,405 +10,70 @@ as $(\text{Skill of last person}) - (\text{Skill of first person})$. Let $s_i$
...
+		    ans %=
+		    M;  // dp[n][0][i] -> n people, 0 unfinished groups, penalty <= x
```
Comment (SansPapyrus683) [inline solutions/gold/cses-1665.mdx:76]: please don't try to squeeze multiple statements in w/ a comma

File: solutions/gold/cses-1665.mdx [hunk 8]
```diff
@@ -10,405 +10,70 @@ as $(\text{Skill of last person}) - (\text{Skill of first person})$. Let $s_i$
...
+	cout << ans;
```
Comment (SansPapyrus683) [inline solutions/gold/cses-1665.mdx:77]:
```suggestion
	cout << ans << endl;
```

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]

File: solutions/gold/cses-1665.mdx (hunk 5 above)
Comment (devo1ution) [reply solutions/gold/cses-1665.mdx:46]: why vectors?

Comment (SansPapyrus683) [reply solutions/gold/cses-1665.mdx]: easier to debug

File: solutions/gold/cses-1665.mdx [hunk 9]
```diff
@@ -2,413 +2,93 @@
...
- * https://codeforces.com/contest/1264/submission/66344993 (ksun)
- * also see https://github.com/ecnerwala/cp-book/blob/master/src/modnum.hpp
- * (ecnerwal) Verification: https://open.kattis.com/problems/modulararithmetic
+/*
+ * dp[N][N][X] -> dp[i][j][k] = first i people, j unfinished groups, k penalty
+ * for convenience, we 1-index dp array and 0-index person array so dp[i][j][k]
+ * lines up with s[i] (i.e. skill of person i + 1 = s[i])
  */
+ll dp1[N][X + K];  // i
+lll dp2[N][X + K];  // i + 1
```
Comment (SansPapyrus683) [inline solutions/gold/cses-1665.mdx]: the comment doesn't explain why there's 2 dp arrays

File: solutions/gold/cses-1665.mdx [hunk 10]
```diff
@@ -2,413 +2,93 @@
...
+		for (int j = 0; j <= (n - i); j++) {  // at most n - i unfinished groups
+			for (int k = 0; k <= x + K; k++) {
+				if (!dp1[j][k]) continue;
+				dp2[j][k] += dp1[j][k];  // i has their own group
+				if (j) dp2[j - 1][k + s[i]] += j * dp1[j][k];  // finish group
+				if (j + 1 <= n - (i + 1))
+					dp2[j + 1][k - s[i]] +=
+					    dp1[j][k];  // create new unfinished group
+				if (j <= n - (i + 1))
+					dp2[j][k] += j * dp1[j][k];  // extend unfinished group
```
Comment (SansPapyrus683) [inline solutions/gold/cses-1665.mdx:73]: braces here & everywhere elsle

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]

File: solutions/gold/cses-1665.mdx [hunk 11]
```diff
@@ -35,42 +35,42 @@ information and problems.
...
+// lines up with s[i] (i.e. skill of person i + 1 = s[i])
+ll dp1[N][X + K]; // dp[i][j][k]
+ll dp2[N][X + K]; // dp[i + 1][j][k]
+// Naively, our dp table would require N * N * X memory
+// However, dp[i][j][k] can only affect dp[i + 1][x][y],
+// so we effectively only need to store two rows of the dp table.
 int main() {
 	cin >> n >> x;
-	for (int i = 0; i < n; i++) { cin >> s[i]; }
+	for (int i = 0; i < n; i++) cin >> s[i];
```
Comment (SansPapyrus683) [inline solutions/gold/cses-1665.mdx]: ayo why remove the braces here

Comment (devo1ution) [reply solutions/gold/cses-1665.mdx]: haha wait that's mb

File: solutions/gold/cses-1665.mdx [hunk 12]
```diff
@@ -2,413 +2,96 @@
...
+int n, x, s[N];     // x -> max penalty
+
+// dp[N][N][X] -> dp[i][j][k] = first i people, j unfinished groups, k penalty
+// for convenience, we 1-index dp array and 0-index person array so dp[i][j][k]
+// lines up with s[i] (i.e. skill of person i + 1 = s[i])
+ll dp1[N][X + K];  // dp[i][j][k]
+ll dp2[N][X + K];  // dp[i + 1][j][k]
+// Naively, our dp table would require N * N * X memory
+// However, dp[i][j][k] can only affect dp[i + 1][x][y],
+// so we effectively only need to store two rows of the dp table.
```
Comment (SansPapyrus683) [inline solutions/gold/cses-1665.mdx:55]:
```suggestion
// Notice we only need to store 2 rows!
ll dp1[N][X + K];  // The subarray we're currently at
ll dp2[N][X + K];  // The next subarray (we'll fill this up with dp1)
```

Comment (SansPapyrus683) [reply solutions/gold/cses-1665.mdx:55]: i hope my understanding of your code is correct

Comment (devo1ution) [reply solutions/gold/cses-1665.mdx:55]: yeah it is :)

File: solutions/gold/cses-1665.mdx (hunk 5 above)
Comment (SansPapyrus683) [reply solutions/gold/cses-1665.mdx]: bump

Comment (devo1ution) [reply solutions/gold/cses-1665.mdx]: i updated it to use vectors, but i dont use them very often so im not sure if there's any standard practices im neglecting

**Review by SansPapyrus683 — APPROVED** [review]

File: solutions/gold/cses-1665.mdx (hunk 5 above)
Comment (SansPapyrus683) [reply solutions/gold/cses-1665.mdx]: nah u did good

**Comment by devo1ution** [comment]
@SansPapyrus683 im pretty new to this; anything else i need to do before this pr gets merged?

**Review by envyaims — APPROVED** [review]
clean uwu

**Pull request merged**